src/main/java/sysc4806/graduateAdmissions/model/Operation.java

```diff
@@ -0,0 +1,22 @@
+package sysc4806.graduateAdmissions.model;
+
+import lombok.Getter;
+
+/**
+ * This enumeration lists possible system operations to be used in the creation of privileges.
+ *
+ * @author luke
+ */
+public enum Operation {
+    CREATE("create"),
+    READ("read"),
+    UPDATE("update"),
+    DELETE("delete");
```

Comment:
I don't understand most of what this comment says, but this is a model object, so there is no "next layer down". These are going to be the actions that a role can do on another entity, and has nothing to do with making REST requests. 

Think along the lines of:
* a Student can create, read, update, and delete their own applications
* admin can read any applications (any probably update them them in case people make mistakes)
* profs can read any application
* students can create, read, update, delete their own interests
* profs can create, read, update, delete their own interests
* admin can read all student and all profs interests

Checking whether or not a user has privileges is going to be something done in the back end, probably done with aspects

Reply:
I think we're talking about the same thing. Users (students, professors, admins) can GET, POST, PUT, DELETE, OPTIONS, etc... resources. Programmers working on the code take this request, check if they have permission to do this, then the CRUD operations are issued. So a POST will be followed by a create cycle one layer down and perhaps other crud operations. Part of the create user cycle, for example, may be an update cycle that updates an invitation status to accepted and a delete cycle that deletes an old account associated with the new account's email, all from the initial POST. 

More about the layers in full stack development:
A stack will have the DOM on the top layer (html/js front end), then the (back end) controllers with endpoints in them in the next layer that the users interact with using REST operations, and then the controllers talk to the JPA layer underneath through CRUD operations which talks to the DB at the bottom of the stack. Basically the same thing, but in terms of permissions, we create permissions on the endpoint and not on database operations.

If this enum is only for database operations, then this is great, but we'll still need a way to control what endpoints each user is allowed to hit so we'll need more verbs to distinguish PUT and PATCH since UPDATE in CRUD can't cover both verbs.
This might help https://www.restapitutorial.com/lessons/httpmethods.html
[More reading about HTTP/REST and CRUD verbs](https://www.google.com/search?q=http+methods+rest+vs+crud+operations&rlz=1C1CHBF_enCA862CA862&oq=http+methods+rest+vs+CRUD)

I've seen Apache Shiro be used as a permissions plugin for spring boot apps. Eventually we could have an endpoint in a UsersController like:
```
@RequestMapping(method = RequestMethod.GET, path="/users/{id}")
@RequiresPermissions(USERS_GET_PERMISSION) /*<--- this permission string will be built using the tools in this pull request */
public ResponseEntity getUser(@PathVariable String id) {
    return ResponseEntity.status(HttpStatus.OK).body(accessDB.getMyUserEntity(id));
}
```
I'm not sure about aspects so I'd love to see an example.

Reply:
It would also be a good idea to use aspects over something else because they are course material, making it good practice for the exam and it shows the prof/TAs that we understand the course material over just doing our own thing.

In the ApplicationController, there would be a method of something like:
``` java
@Controller
@EnableAspectJAutoProxy
public class ApplicationController{

    ...

    @PostMapping("/add-app")
    @ResponseBody
    public Application createApplication(
                @RequestBody Application app, @RequestBody userID){
        return repo.save(app);
    }
}
```

for this, the aspect class would look like:
``` java
@Aspect
public class ApplicationAspect{
    @Around("com.myapp.controller.ApplicationController.createApplication() && 
                args(application, userID)")
    public Object checkPrivilege(ProceedingJoinPoint pjp, 
                Application application, long userID){
        if(/*specified user has access privileges*/)
            return pjp.proceed();
        else
            return /*some default access denied info*/
    }
}
```

This may not be exactly correct, since i'm new to using AOP and I wrote this code just directly into this comment, but thats the gist of how it would look. This specific example stops a user adding an application to the data base. A similar type of method can be written to decide if the user should get the form to create the application in the first place.

If your issue is that the operations follow CRUD names, that is partially just because these names have an obvious meaning, whereas HTTP request types do not. For an admin (or a 4th as yet undefined role) that can edit role privileges, GET, POST, etc. only make sense to some one with web dev knowledge, whereas CRUD operations have an obvious meaning behind the names. To just use the system, you shouldn't need any specialized IT knowledge. I could've just as easily named CREATE as "add" instead, or name DELETE as "remove", these names aren't limiting the use of these roles to only methods in the system that directly do CRUD things.

If the problem is just the names included in the enum. I would suggest we merge this and make what the operations should be a point of discussion next week in the meeting. Roles need this object to be created, Users need a Role to be create, and Applications need a User to be created, so until this object is ready, everything else is halted.

Reply:
Here's another way to think of this that might make more sense.

To create/add a new graduate school application, you're going to have to do two things:
1. GET the form to fill out for the application
2. POST your information so the application can be added to the database

It would be silly to give a role privileges to do one, but not the other, any role that has one should get the other. This is also true for updating an application, or adding/updating interests, terms, etc.

If you gave people individual privileges that were for GET, POST, etc. you're going to end up with privileges that don't make sense to be a standalone thing, and you're also going to have a Privilege table that has way more records.

Reply:
Ah okay. As long as the names are cosmetic for an end user, that's fine. I'd assume we're going to give admins the ability to alter specific permissions. Aspects seem okay too. It sounded like we were going to put permissions on database operations and not endpoints but your example cleared that up, so that's good.

Addressing the comment above, I don't think I explained what I was talking about properly because I completely agree. I can provide a working example another time maybe.